### ISSUE TITLE
IS this dead? And is this client or server code?

### ISSUE THREAD

**Issue by Colonizor48** [rafern/diep.io-wasm-re#1]

**Comment by rafern** [comment]
It's client code. I started it for fun and then I realised that this project is torture because the decompiled code is so bad and it changes every update, so I kinda just gave up. Maybe I'll return to it later, but probably not

**Comment by Colonizor48** [comment]
I want to figure out a way to remake arras.io with the origional diep.io engine. Getting the physics of diep.io and the new mechanics of arras.io

**Comment by Colonizor48** [comment]
Or just figure out a way to get the server source code for the current version of arras.io

**Comment by Colonizor48** [comment]
But I am 99% sure this is diep.io's last update. Even if it isn't the ones after are going to be minor. I propose we should try to figure out a way to either get the server source code directly or remake a custom server software with a way to connect to it in diep.io

**Comment by Colonizor48** [comment]
I would make a new repo and a discord for unifying everything into 1. But i have no idea how any of this works(i want to learn tho and i am familiar with very basic c++) And i'm not allowed to use discord. So that means having to rely on steam groups or something smaller.

**Comment by Colonizor48** [comment]
allright i'll post all of the repos i can find here.
1. https://github.com/ajsdev/deip.oi it seems that there was work on a private server as early as 2017. Some things could probably be salvaged from here
2. https://github.com/HueHanaejistla/diep.io-protocol-wiki, https://github.com/Faris90/Diep.io-Protocol, https://github.com/firebolt55439/Diep.io-Protocol/issues Various sources documenting the protocol of diep.io Possibly outdated and possibly conflicting info. Will have to be tested
3. https://github.com/YoussefMrabet/diepio.github.io/blob/gh-pages/index.html ancient 2015 code. Could be useful idk.
4. https://github.com/cx88/diepssect. By far the most useful thing. A metric tone of info. Including many(outdated) versions. But a treasure trove of scripts and knowlage. This is where that image of a player controled guardian comes from. And what i hope we could eventually recreate.
5. I will post more useful repos as i find them.

**Comment by Colonizor48** [comment]
I want to try to help you. I have no idea how but i want to.

**Comment by Colonizor48** [comment]
Also there is a bit of activity in the diepssect issues section. Someone want sot try to revive it in disord. I cant use discord but still if you are interasted reach out. Sorry for the thread spam.

**Comment by rafern** [comment]
To be honest, you'll probably get better luck by checking with other projects instead of this one, since I'm not taking this as seriously as you (again it was just for fun and I started to find it tedious) and even if I did start finding it fun again I haven't had enough free time recently. I guess all I can do is explain to you why its so tedious to reverse engineer the code:
- The wasm2c decompiler is horrible. WASM works as a stack machine, so, for example, `x = 5 * 6 + 7` gets converted to `push 5, push 6, multiply, push 7, add`. Instead of wasm2c converting generated WASM code back to the more readable (I guess you would call it canonical/standard?) code, it converts it back to something like `temp_1 = 5; temp_2 = 6, temp_3 = temp_1 * temp_2; temp_4 = 7; temp_5 = temp_3 + temp_4` instead of just `temp_1 = 5 * 6 + 7`. Basically, the decompiled code feels like reading assembly, which is almost no better than me just looking at the WASM code.
- Even though there haven't been any content updates to the game, the game has been silently updated at least on the client side multiple times, without anything in the changelog saying what changed. I have no idea what the updates added, but they changed all the function mappings, basically throwing all of my progress in the trash. I don't know if this is still the case, but at the time I was working on this project, it was

The first problem could be fixed by either patching wasm2c to stop being so terrible, or to write a program that turns the decompiled C code generated by wasm2c into an intermediate language (like LLVM IR; there might already be a way to do this with the LLVM toolchain), then turn the intermediate code into a canonical form (without all those unnecessary temporary variables), then revert back to C. For both of those, you need a lot of time and effort.

The second problem can be fixed by comparing old decompiled functions to new ones and updating the mappings (what the obfuscated functions names are, like how minecraft forge/fabric does it) which is a lot easier than the first problem, but to do that, you first need to fix the first problem.

Even after solving those problems, the decompiled code is C, but the original code is written in C++, so you'll get the mangled functions instead of classes. There is no way to decompile C++ code so that classes are preserved, because unlike Java, C++ converts all class methods to normal functions, essentially deleting all information that a function is actually a class method, unless you have debug symbols enabled (the game doesn't, since its WASM). For example, if a there is a class that has a bunch of methods, you'll get, in the decompiled code, a bunch of random seemingly unrelated functions; multiple functions for the different constructors for that class (copy constructor, move constructor, default contructor, etc...), the class destructor, and then the actual methods converted to functions. Basically, you have to guess which functions are actually class methods, and then also guess if they belong to the same class. Even worse, you also have to guess which functions are part of the standard C++ library (the game uses emscripten, which uses LLVM's libcxx and some functions from glibc) by comparing your decompiled code with the source code from the standard library (https://github.com/llvm-mirror/libcxx and https://github.com/bminor/glibc).

In the end, you need a lot of time and motivation for this project, which I don't have anymore. I might return to it in the future, but for now you can consider this project in hiatus or just dead.

TLDR (sorry for the wall of text): the wasm2c decompiler is terrible and makes it tedious to reverse engineer, so if you want to find out how diep.io physics work, you will probably have better luck by checking other reverse engineering projects with people that are still motivated. Also, thanks for the links

**Comment by diepiodiscord** [comment]
> I saw a picture from 2 years ago of someone playing as a guardian.

That screenshot was from M28 himself (the dev) showing off

**Comment by rafern** [comment]
Also, if you want game physics, that is most likely server-sided. This project is for client side only, since you can't legally get the diep.io server, unless the developer included both client and server in the same binary (like the orbs.it developer did), but I don't think that's the case
